src/timesync/timesyncd-server.c

```diff
@@ -175,3 +178,39 @@ void server_name_flush_addresses(ServerName *n) {
         while (n->addresses)
                 server_address_free(n->addresses);
 }
+
+int server_name_parse_port(ServerName *n) {
+        const char *sqr = "]", *col = ":";
+        char *ret_sqr, *ret_col, *last_col;
+
+        ret_sqr = strrchr(n->string, *sqr);
+        ret_col = strchr(n->string, *col);
+        last_col = strrchr(n->string, *col);
+
+        if (ret_sqr == NULL && ret_col == NULL) { // server.domain or I.P.v.4
+                return 0;
+        } else if (ret_sqr == NULL && strlen(ret_col) == strlen(last_col)) { // has no ']' and exactly one ":"
+                const char *word = strdupa_safe(n->string);
+                (void) extract_first_word(&word, &n->string, col, 0);
+                (void) extract_first_word(&word, &n->overridden_port, col, 0);
```

Comment:
i really don't like that n->string changes contents once server_name_parse_port() is invoked.

we want that our fields are reasonably normalized, they contain the same type of data all the time.

the port number should already be parsed the moment we read it from the config file, and never be maintained in server_name_new() as part of the server string, it should be split off there already.

Reply:
The lifecycle runs in such a way that `timesyncd-conf.c` has `manager_parse_server_string()` which calls `timesyncd-server.c`'s `server_name_new()` to create the `ServerName` struct. Then the `ServerAddress` struct where you might store the port number is not created until `manager_connect()`  calls resolve_getaddrinfo() with `m->current_server_name->string` as the server address.

Consequently, I'll move this so that `server_name_new()` doesn't do anything (even removing the changes to the `ServerAddress` struct), then using `manager_connect()` in `timesyncd-manager.c` to parse the `m->current_server_name->string` to see if it has port information which we can use appropriately. (It's a better match to the manager using `server_name_flush_addresses()` just before calling `resolve_getaddr_info()`).